iree/compiler/Conversion/LinalgToSPIRV/FlattenMemRefSubspanPass.cpp

```diff
@@ -0,0 +1,342 @@
+  using OpConversionPattern<memref::LoadOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      memref::LoadOp loadOp, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
+    memref::LoadOp::Adaptor adaptor(operands);
+    if (!isRankZeroOrOneMemRef(adaptor.memref().getType())) {
+      return rewriter.notifyMatchFailure(
+          loadOp, "expected converted memref of rank <= 1");
+    }
+
+    Value linearIndex = linearizeIndices(
+        loadOp.getMemRefType(), loadOp.getIndices(), loadOp.getLoc(), rewriter);
+    rewriter.replaceOpWithNewOp<memref::LoadOp>(loadOp, adaptor.memref(),
+                                                linearIndex);
+    return success();
+  }
+};
+
+/// Linearizes indices in memref.store ops.
+struct LinearizeStoreIndices final
+    : public OpConversionPattern<memref::StoreOp> {
+  using OpConversionPattern<memref::StoreOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      memref::StoreOp storeOp, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const override {
+    memref::StoreOp::Adaptor adaptor(operands);
+    if (!isRankZeroOrOneMemRef(adaptor.memref().getType())) {
+      return rewriter.notifyMatchFailure(
+          storeOp, "expected converted memref of rank <= 1");
+    }
+
+    Value linearIndex =
+        linearizeIndices(storeOp.getMemRefType(), storeOp.getIndices(),
+                         storeOp.getLoc(), rewriter);
+    rewriter.replaceOpWithNewOp<memref::StoreOp>(storeOp, adaptor.value(),
+                                                 adaptor.memref(), linearIndex);
+    return success();
+  }
+};
+
+//===----------------------------------------------------------------------===//
+// Folding Patterns
+//===----------------------------------------------------------------------===//
+
+/// Returns the number of bytes of the given `type`. Returns llvm::None if
+/// cannot deduce.
+///
+/// Note that this should be kept consistent with how the byte offset was
+/// calculated in the subspan ops!
+Optional<int64_t> getNumBytes(Type type) {
+  if (type.isIntOrFloat()) return (type.getIntOrFloatBitWidth() + 7) / 8;
+  if (auto vectorType = type.dyn_cast<VectorType>()) {
+    auto elementBytes = getNumBytes(vectorType.getElementType());
+    if (!elementBytes) return llvm::None;
+    return elementBytes.getValue() * vectorType.getNumElements();
+  }
+  return llvm::None;
+}
+
+/// Folds the byte offset on subspan ops into the consumer load/store ops.
+template <typename OpType>
+struct FoldSubspanOffsetIntoLoadStore final : public OpRewritePattern<OpType> {
+  using OpRewritePattern<OpType>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(OpType op,
+                                PatternRewriter &rewriter) const override {
+    auto memrefType = op.memref().getType().template cast<MemRefType>();
+    if (!isRankZeroOrOneMemRef(memrefType)) return failure();
```

Comment:
Nit: To be consistent, the similar check in the load returns a `notifyMatchFailure`. Maybe do the same here.